components/raftstore/src/store/fsm/peer.rs

```diff
@@ -1974,53 +2019,30 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             return Ok(true);
         }
 
-        let state_key = keys::region_state_key(region_id);
-        let state: RegionLocalState = match self.ctx.engines.kv.get_msg_cf(CF_RAFT, &state_key) {
+        // All of the target peers must exist before merging which is guaranteed by PD.
+        // Now the target peer is not in region map.
+        match self.is_merge_target_region_stale(target_region) {
             Err(e) => {
                 error!(
                     "failed to load region state, ignore";
                     "region_id" => self.fsm.region_id(),
                     "peer_id" => self.fsm.peer_id(),
                     "err" => %e,
-                    "target_region_id" => region_id,
+                    "target_region_id" => target_region_id,
                 );
-                return Ok(false);
+                Ok(false)
             }
-            Ok(None) => {
-                info!(
-                    "seems to merge into a new replica of region, let's wait.";
-                    "region_id" => self.fsm.region_id(),
-                    "peer_id" => self.fsm.peer_id(),
-                    "target_region_id" => region_id,
-                );
-                return Ok(false);
+            Ok(true) => Err(box_err!("region {} is destroyed", target_region_id)),
```

Comment:
This is not correct. Target peer doesn't have to be destroyed to reach this branch.

Reply:
If the target peer is not destroyed, it should in region map, but it's not. Even if PD does not ensure all target peer exists, this source peer should also consider rollback merge because the local target peer is fresher.